lib/graphql/sugar/object.rb

```diff
@@ -1,21 +1,37 @@
 module GraphQL
   module Sugar
     module Object
+      COMMON_ATTRIBUTES = [:id, :created_at, :updated_at]
+
       def self.included(base)
         base.extend ClassMethods
       end
 
       module ClassMethods
+        def inherited(klass)
+          derived_model_class = klass.name.demodulize.gsub('Type', '').safe_constantize
+          klass.instance_variable_set(:@model_class, derived_model_class)
+          klass.instance_variable_set(:@attributes_initialized, Set.new)
+        end
+
         def model_class(model_class)
           @model_class = model_class
+          initialize_common_attributes unless all_initial_attributes_added?
+        end
 
-          common_attribute_names = [:id, :created_at, :updated_at]
-          common_attribute_names.each(&method(:attribute))
+        def initialize_common_attributes
+          COMMON_ATTRIBUTES.each do |_attribute|
+            @attributes_initialized.add(_attribute)
+            attribute(_attribute)
+          end
         end
 
         def attribute(name, return_type_expr = nil, desc = nil, **kwargs, &block)
           raise "You must define a `model_class` first in `#{self.name}`." if @model_class.blank?
 
+          initialize_common_attributes if @attributes_initialized.exclude?(name) && !all_initial_attributes_added?
+          @attributes_initialized.add(name)
```

Comment:
1. I would simplify to:

```
return if @attributes_initialized.include?(name)
@attributes_initialized.add(name)
```

2. Since model_class is no longer required, we need to call initialize_common_attributes in both `attribute` and `model_class`.

```
initialize_common_attributes
```

3. Remove the `all_initial_attributes_added?` method as the check would be present within the `attribute` method itself.

Reply:
1. the above check for the `@attributes_initialized` is to prevent the recursion and would create the duplicate types. This is just the mechanism to control the recursion when we want to run the `intialize_common_attributes`.

2. We still need the `model_class` method in those scenarios when the Automatic type detection failed. See: https://github.com/Amandeepsinghghai/sample-graphql-app/blob/tmp-work/app/graphql/types/optional_user_type.rb#L3

3. I am all ears, on this, but I also felt that this can improved.